### DISCUSSION TITLE
How to format data set to model duration of 2-comp oral dosing

### DISCUSSION THREAD

**Discussion by Dstix** [nlmixr2/nlmixr2#186] [Q&A] [answered]
I am trying to model a 2-compartment model with duration modeled:
```r
model_2cpt_ode <- function() {
  ini({
    tka <- log(0.2)
    tcl <- log(23)
    tv2  <- log(8)
    tv3  <- log(90)
    tq   <- log(1)
    d1 <- log(2)
    eta.ka ~ 0.1
    eta.cl ~ 0.1
    eta.d1 ~ 0.1
    prop.err <- 0.4
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v2 <- exp(tv2)
    v3 <- exp(tv3)
    q  <- exp(tq)
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v2 * center + q/v3 * periph - q/v2 * center
    d/dt(periph) = q/v2 * center - q/v3 * periph
    dur(depot) = exp(d1 + eta.d1)
    cp = 1000 * center / v2 
    cp ~ prop(prop.err)
  })
}
```

The trouble is, I am not certain how to specify this in the data set. Should the row with `AMT` be `RATE = -2` and `EVID = 101`? If I do that, I get `'rate' or 'dur' is ignored with classic rxode2 'EVID's` and it is clearly not solving for d1.

If I code it as `RATE = -2` and `EVID = 1` it seems to work, but the results are unconvincing. I am not certain I am dosing the right compartment. I put in a `CMT = 1` for the dosing row and `CMT = 2` for the observation row, but I am still not certain. 

I have looked at other discussions, including #154 but it doesn't seem to answer my data set question.

**Comment by mattfidler** [comment, marked as answer]
In general, you should use `-2` and `AMT`  with the `CMT` specification to specify an infusion with a modeled duration. 

Note that you are currently infusing to the depot, and that may be why you are getting the results that are not that convincing.  Most infusions occur to the central compartment.

Here is an annotated example:

``` r
library(rxode2)
#> rxode2 2.0.13.9000 using 4 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`

model_2cpt_ode <- function() {
  ini({
    tka <- log(0.2)
    tcl <- log(23)
    tv2  <- log(8)
    tv3  <- log(90)
    tq   <- log(1)
    d1 <- log(2)
    eta.ka ~ 0.1
    eta.cl ~ 0.1
    eta.d1 ~ 0.1
    prop.err <- 0.4
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v2 <- exp(tv2)
    v3 <- exp(tv3)
    q  <- exp(tq)
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v2 * center + q/v3 * periph - q/v2 * center
    d/dt(periph) = q/v2 * center - q/v3 * periph
    dur(depot) = exp(d1 + eta.d1)
    cp = 1000 * center / v2 
    cp ~ prop(prop.err)
  })
}

f <- model_2cpt_ode()

# This is the underlying rxode2 classic model
summary(f$simulationModel)
#> rxode2 2.0.13.9000 model named rx_f1020d3296ac324ece743fff4fb03582_x6 model (✔ ready). 
#> DLL: C:\Users\fidlema3\AppData\Local\Temp\1\RTMPUS~1\rxode2\RX_F10~1.RXD/rx_f1020d3296ac324ece743fff4fb03582_x64.dll
#> NULL
#> 
#> Calculated Variables:
#> [1] "ka"       "cl"       "v2"       "v3"       "q"        "cp"       "ipredSim"
#> [8] "sim"     
#> ── rxode2 Model Syntax ──
#> rxode2({
#>     param(tka, tcl, tv2, tv3, tq, d1, prop.err, eta.ka, eta.cl, 
#>         eta.d1)
#>     ka = exp(tka + eta.ka)
#>     cl = exp(tcl + eta.cl)
#>     v2 = exp(tv2)
#>     v3 = exp(tv3)
#>     q = exp(tq)
#>     d/dt(depot) = -ka * depot
#>     d/dt(center) = ka * depot - cl/v2 * center + q/v3 * periph - 
#>         q/v2 * center
#>     d/dt(periph) = q/v2 * center - q/v3 * periph
#>     dur(depot) = exp(d1 + eta.d1)
#>     cp = 1000 * center/v2
#>     rx_yj_ ~ 2
#>     rx_lambda_ ~ 1
#>     rx_low_ ~ 0
#>     rx_hi_ ~ 1
#>     rx_pred_f_ ~ cp
#>     rx_pred_ ~ rx_pred_f_
#>     rx_r_ ~ (rx_pred_f_ * prop.err)^2
#>     ipredSim = rxTBSi(rx_pred_, rx_lambda_, rx_yj_, rx_low_, 
#>         rx_hi_)
#>     sim = rxTBSi(rx_pred_ + sqrt(rx_r_) * rxerr.cp, rx_lambda_, 
#>         rx_yj_, rx_low_, rx_hi_)
#>     cmt(cp)
#>     dvid(4)
#> })

# You can see that ipredSim shows the indiviaul value, and sim includes residuals

e <- et(amt=1000, rate=-2) %>%
  et(0, 4, length.out=20) %>%
  et(0, 24, length.out=50)

# You can see that the duration with -2 is a modled infusion
e
#> ── EventTable with 71 records ──
#> 1 dosing records (see x$get.dosing(); add with add.dosing or et)
#> 70 observation times (see x$get.sampling(); add with add.sampling or et)
#> ── First part of x: ──
#> # A tibble: 71 × 4
#>     time   amt rate       evid         
#>    <dbl> <dbl> <rate/dur> <evid>       
#>  1 0        NA NA         0:Observation
#>  2 0        NA NA         0:Observation
#>  3 0      1000 -2:dur     1:Dose (Add) 
#>  4 0.211    NA NA         0:Observation
#>  5 0.421    NA NA         0:Observation
#>  6 0.490    NA NA         0:Observation
#>  7 0.632    NA NA         0:Observation
#>  8 0.842    NA NA         0:Observation
#>  9 0.980    NA NA         0:Observation
#> 10 1.05     NA NA         0:Observation
#> # ℹ 61 more rows

s <- rxSolve(model_2cpt_ode, e)

# this plot isn't that convincing of a 2 hour duration infusion
library(patchwork)
plot(s, ipredSim, log="y") * plot(s, ipredSim)
#> Warning: Transformation introduced infinite values in continuous y-axis
```

![](https://i.imgur.com/HqfknOi.png)

``` r

# But if you extend it out for say 9 hours it is more noticble

s <- f %>% ini(d1=log(14)) %>% rxSolve(e)
#> ℹ change initial estimate of `d1` to `2.63905732961526`

plot(s, ipredSim, log="y") * plot(s, ipredSim)
#> Warning: Transformation introduced infinite values in continuous y-axis
```

![](https://i.imgur.com/DkIMSzz.png)

``` r

# and even more notiable if the infusion is in the central compartment

f2 <- f %>% 
  model(-dur(depot)) %>% 
  model(dur(center)<-exp(d1+eta.d1)) %>%
  ini(d1=log(9), eta.d1 ~ 0.1)
#> ! remove population parameter `d1`
#> ! remove between subject variability `eta.d1`
#> ℹ add covariate `d1`
#> ℹ add between subject variability `eta.d1` and set estimate to 1
#> ℹ promote `d1` to population parameter with initial estimate 2.19722457733622
#> ℹ change initial estimate of `d1` to `2.19722457733622`
#> ℹ change initial estimate of `eta.d1` to `0.1`

e2 <- et(amt=1000, rate=-2, cmt="center") %>%
  et(0, 4, length.out=20) %>%
  et(0, 24, length.out=50)

s <- f2 %>% ini(d1=log(9)) %>% rxSolve(e2)
#> ℹ change initial estimate of `d1` to `2.19722457733622`

plot(s, ipredSim, log="y") * plot(s, ipredSim)
#> Warning: Transformation introduced infinite values in continuous y-axis
```

![](https://i.imgur.com/aKrkSKV.png)

``` r

# The way to code it using the other form of events is using the flags 8 and 6
# since you are dosing to the central compartment (compartment #2) you would
# use the base of 201 and then prepend 80 for when the duration turns on and 60 
# for when the duration.  In `rxode2` these have to be right next to each other
# because the calculated duration/rate is stored in the next record.
# 
# This means the dose would be:
#
# TIME EVID  AMT
#    0 80201 1000
#    0 60201 1000
# 
# no matter which approach is used this is evetually what is seen by the solver
#
# YOu can check the underlying translation by:

head(etTrans(e2, f2$simulationModel))
#>   ID      TIME  EVID  AMT II DV
#> 1  1 0.0000000 80201 1000  0 NA
#> 2  1 0.0000000 60201 1000  0 NA
#> 3  1 0.0000000     0   NA  0 NA
#> 4  1 0.0000000     0   NA  0 NA
#> 5  1 0.2105263     0   NA  0 NA
#> 6  1 0.4210526     0   NA  0 NA
#> 
#> Covariates (non time-varying):
#>   ID
#> 1  1
#> 
#> Compartment translation:
#>   Compartment Name Compartment Number
#> 1            depot                  1
#> 2           center                  2
#> 3           periph                  3
#> 4               cp                  4
```

<sup>Created on 2023-09-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by Dstix** [comment]
Super helpful Matt, thanks!

My goal is to model the case where the entire dose is not immediately available in the gut (e.g., the tablet is still dissolving). So I don't expect it to be very dramatic, but I am trying capture different absorption shapes - putting it into the central compartment wouldn't apply in my case. 

_For what it is worth, when I switch from focei to saem, the model seems to "click" into place better._ Not sure why, but this is not the first time. I find I don't have to strong-arm the saem model as much as focei...

As a follow up: If I want to specify the CMT where the RATE = -2 goes, to specify it in the nlmixr2 input data set as `CMT = 1` or `CMT = "depot"`? Or would both work? That is:

```
pk_nl %<>% tibble(
  ID = 1,
  TIME = 0,
  AMT = 1000,
  DV = 0,
  RATE = -2,
  CMT = "depot"
)
```

**Reply by mattfidler** [reply]
> My goal is to model the case where the entire dose is not immediately available in the gut (e.g., the tablet is still dissolving). So I don't expect it to be very dramatic, but I am trying capture different absorption shapes - putting it into the central compartment wouldn't apply in my case.

While I believe this is true, some times it is difficult to distinguish from a simple first order absorption.  You can see why in the plots above.

> For what it is worth, when I switch from focei to saem, the model seems to "click" into place better. Not sure why, but this is not the first time. I find I don't have to strong-arm the saem model as much as focei...

In general the `SAEM` doesn't use gradients and also uses the full likelihood and is often more stable.  That being said, it is also more likely to give you unidentifiable solutions (whereas `focei` is likely to choke on unidentifiable solutions).  Many people think that `focei` will keep the solutions more "honest" though I would say whatever works is best to use.

In this case you are making a more difficult to estimate problem by having a zero order process followed by a first order absorption.  Depending on the data it may be unidentifiable and this is why saem is performing better than focei.

Also saem and focei may perform differently on different problems.  `saem` prefers problems with many `eta`s and `focei` prefers problems with few `eta`s.  

> As a follow up: If I want to specify the CMT where the RATE = -2 goes, to specify it in the nlmixr2 input data set as CMT = 1 or CMT = "depot"? Or would both work? That is:

Both would work.  I prefer `"depot"` since you can change the model as long as the compartment name remains the same you don't have to change your input data.   That being said you need to use `CMT` as a number to be able to use the same dataset with `NONMEM`
